### PR TITLE
Purchases: Add `CancelPurchasePlaceholder`

### LINKED ISSUE #275: Purchases: Add loading page skeleton for Single Purchase pages
We need design for Purchase pages loading state. This includes following pages:
- Manage Purchase
- Cancel Purchase
- Confirmation Cancel Purchase
- Edit Payment Method (moved to #843)
- Edit Card Details

As part of this task we should also add redirects for all aforementioned pages when there is no purchase selected.

### Loading Placeholders
@mikeshelton1503 proposed the following mockups:

* [x] Manage Purchase
screen shot 2015-10-21 at 1 44 21 pm

* [x] [Cancel Purchase](https://github.com/Automattic/wp-calypso/pull/915)
screen shot 2015-10-21 at 1 44 29 pm

* [x] [Confirm Cancel Purchase](https://github.com/Automattic/wp-calypso/pull/1036)
screen shot 2015-10-21 at 1 44 37 pm

* [x] [Edit Card Details](https://github.com/Automattic/wp-calypso/pull/717)
screen shot 2015-10-21 at 1 44 52 pm

### Page Redirects
* [x] [Perform redirect when user can't access Manage Purchase page](https://github.com/Automattic/wp-calypso/pull/1255)
* [x] [Perform redirect when user can't access Cancel Purchase page](https://github.com/Automattic/wp-calypso/pull/1153)
* [ ] Perform redirect when user can't access Confirm Cancel Purchase page
* [ ] Perform redirect when user can't access Cancel Private Registration page
* [ ] Perform redirect when user can't access Edit Card Details page

### Refactoring tasks
* [x] ~~Create Edit Card Details skeleton that can be shared by loading placeholder and page component~~ - it produced more lines of code, than using duplicated code in case of Edit Card Details page, let's skip it for the time being
* [x] [Move all page titles into one place](https://github.com/Automattic/wp-calypso/pull/1030)
* [x] [Always use selectedSite.slug in urls when referencing site domain](https://github.com/Automattic/wp-calypso/pull/1147)

### PR DESCRIPTION
**Pull request by drewblaisdell**
This PR adds `CancelPurchasePlaceholder` and updates the cancel purchase page to render this placeholder when purchases are fetching.

I added this component to `me/purchases/cancel-purchase-placeholder/` instead of `me/purchases/cancel-purchase/placeholder.jsx` so that it can be reused as the fetching state of `ConfirmCancelPurchase`, as the placeholder design is identical. `ConfirmCancelPurchase` has two loading phases (fetching `/purchases`, then fetching the endpoint cancel form), and further, non-trivial changes are required to overwrite the 'Loading..' text with `CancelPurchasePlaceholder` on that page.

Fixes #275 (not entirely, but I want to link this to that issue).

**Testing**
- Visit http://calypso.localhost:3000/purchases
- Click on a purchase.
- Click 'Cancel and Refund [purchase]'.
- Refresh the page.
- Assert that the fetching state contains a placeholder like in the design in #275.

- [x] QA review
- [x] Code review

### CODE REVIEW

File: shared/components/data/store-connection/index.jsx [hunk 1]
```diff
@@ -59,7 +59,7 @@ const StoreConnection = React.createClass( {
 
 	render() {
 		if ( this.isDataLoading() ) {
-			return React.createElement( this.props.loadingPlaceholder );
+			return React.createElement( this.props.loadingPlaceholder, this.state );
```
Comment (drewblaisdell) [inline shared/components/data/store-connection/index.jsx]: This PR also contains an update to `StoreConnection` to pass the state of the stores (which contain the given props, which is why the change is necessary) to `LoadingPlaceholder`. cc @gziolo

File: client/components/loading-placeholder/index.jsx [hunk 2]
```diff
@@ -15,7 +15,7 @@ const LoadingPlaceholder = React.createClass( {
 	propTypes: {
 		className: React.PropTypes.string,
 		path: React.PropTypes.string,
-		title: React.PropTypes.string.required
+		title: React.PropTypes.string.isRequired
```
Comment (gziolo) [inline client/components/loading-placeholder/index.jsx]: Thanks for fixing :)

File: shared/components/data/store-connection/index.jsx (hunk 1 above)
Comment (gziolo) [inline shared/components/data/store-connection/index.jsx]: Definitely it's something we need in this case :+1:

File: client/components/data/purchases/manage-purchase/index.jsx [hunk 3]
```diff
@@ -35,7 +35,9 @@ const ManagePurchaseData = React.createClass( {
 		component: React.PropTypes.func.isRequired,
 		purchaseId: React.PropTypes.string.isRequired,
 		sites: React.PropTypes.object.isRequired,
-		destinationType: React.PropTypes.string
+		destinationType: React.PropTypes.string,
+		loadingPlaceholder: React.PropTypes.func,
+		isDataLoading: React.PropTypes.func
```
Comment (gziolo) [inline client/components/data/purchases/manage-purchase/index.jsx]: We could also reference `isDataLoading` directly here. I guess both way works the same ;)

**Comment by gziolo** [comment]
Looks and works as expected. I'll merge it, because I need it to continue on #275 :)
Good work @drewblaisdell :+1:
